tokio-trace-core/src/field.rs

```diff
@@ -30,74 +34,256 @@
+        for (k, v) in i {
+            self.record_kv(k, v)?;
+        }
+        self.close_map()
+    }
 
-    /// Returns some reference to the boxed value if it is of type `T`, or
-    /// `None` if it isn't.
-    fn downcast_ref<T: AsValue + 'static>(&self) -> Option<&T>
+    /// Record an ordered list of `Value`s.
+    ///
+    /// This function manages calling `open_list`, recording the list elements
+    /// in the given iterator, and closing the list.
+    ///
+    /// This is the suggested way for `Value` implementations to record lists,
+    /// rather than calling those functions directly, unless different behaviour
+    /// is needed.
+    pub fn record_list<'a, I, V>(&mut self, i: I) -> RecordResult
     where
-        Self: 'static;
-}
+        I: IntoIterator<Item = &'a V>,
+        V: Value + 'a,
+    {
+        self.open_list()?;
+        for v in i {
+            v.record(self)?;
+        }
+        self.close_list()
+    }
 
-/// Trait implemented by types which may be converted into a `Value`.
-///
-/// Implementors of `AsValue` must provide an implementation of the `fmt_value`
-/// function, which describes how to format a value of this type.
-pub trait AsValue: Send + Sync {
-    /// Formats the value with the given formatter.
-    fn fmt_value(&self, f: &mut fmt::Formatter) -> fmt::Result;
-
-    #[doc(hidden)]
-    fn type_id(&self) -> TypeId
+    /// Record a `struct` list of `Value`s, given the struct's `name` and an
+    /// iterator over its `fields`.
+    ///
+    /// This function manages calling `open_struct`, recording the struct's
+    /// fields, and closing the struct.
+    ///
+    /// This is the suggested way for `Value` implementations to record structs,
+    /// rather than calling those functions directly, unless different behaviour
+    /// is needed.
+    pub fn record_struct<'a, I>(&mut self, name: &str, fields: I) -> RecordResult
     where
-        Self: 'static,
+        I: IntoIterator<Item = (&'a str, &'a dyn Value)>,
     {
-        TypeId::of::<Self>()
+        self.open_struct(name)?;
+        for (name, v) in fields {
+            self.record_kv(&name, v)?;
+        }
+        self.close_struct()
+    }
+
+    /// Record a tuple.
+    ///
+    /// This function manages calling `open_tuple`, recording the tuple's
+    /// fields, and closing the tuple.
+    ///
+    /// This is the suggested way for `Value` implementations to record tuples,
+    /// rather than calling those functions directly, unless different behaviour
+    /// is needed.
+    pub fn record_tuple<'a, I>(&mut self, i: I) -> RecordResult
+    where
+        I: IntoIterator<Item = &'a dyn Value>,
+    {
+        self.open_tuple()?;
+        for v in i {
+            v.record(self)?;
+        }
+        self.close_tuple()
     }
 }
 
-/// Trait representing a type which may be converted into an `OwnedValue`.
+/// Result type returned by recording field values with a `Record`.
+pub type RecordResult = Result<(), Error>;
```

Comment:
I'm a bit skeptical of introducing Result aliases. I can be convinced that this is a necessary change, but I'd like to understand why.

Reply:
The result type alias isn't strictly necessary, I just thought it was nice to reduce the amount I had to type. Remember that the type can still be _implemented_ with fns returning `Result<(), field::Error>`, that's equivalent to the type alias.

IMO result type aliases are only problematic when the name of the type alias is _also_ `Result` --- it's the name collision that's an issue. I'm happy to get rid of this if you have a problem with it, though.

Reply:
> IMO result type aliases are only problematic when the name of the type alias is also Result --- it's the name collision that's an issue. I'm happy to get rid of this if you have a problem with it, though.

That's really fair—I forgot about that bit. I have no problem with the alias, then.